action/protocol/staking/protocol.go

```diff
@@ -798,33 +839,62 @@ func (p *Protocol) needToWriteCandsMap(ctx context.Context, height uint64) bool
 	return height >= p.config.PersistStakingPatchBlock && fCtx.CandCenterHasAlias(height)
 }
 
-func (p *Protocol) contractStakingVotes(ctx context.Context, candidate address.Address, height uint64) (*big.Int, error) {
+func (p *Protocol) contractStakingVotes(ctx context.Context, candidate address.Address, height uint64, view *ViewData) (*big.Int, error) {
 	featureCtx := protocol.MustGetFeatureCtx(ctx)
 	votes := big.NewInt(0)
 	indexers := []ContractStakingIndexer{}
+	views := []ContractStakeView{}
 	if p.contractStakingIndexer != nil && featureCtx.AddContractStakingVotes {
-		indexers = append(indexers, p.contractStakingIndexer)
+		if featureCtx.CreatePostActionStates {
+			views = append(views, view.contractsStake.v1)
+		} else {
+			indexers = append(indexers, p.contractStakingIndexer)
+		}
 	}
 	if p.contractStakingIndexerV2 != nil && !featureCtx.LimitedStakingContract {
-		indexers = append(indexers, p.contractStakingIndexerV2)
+		if featureCtx.CreatePostActionStates {
+			views = append(views, view.contractsStake.v2)
+		} else {
+			indexers = append(indexers, p.contractStakingIndexerV2)
+		}
 	}
 	if p.contractStakingIndexerV3 != nil && featureCtx.TimestampedStakingContract {
-		indexers = append(indexers, p.contractStakingIndexerV3)
+		if featureCtx.CreatePostActionStates {
+			views = append(views, view.contractsStake.v3)
+		} else {
+			indexers = append(indexers, p.contractStakingIndexerV3)
+		}
+	}
+
+	accountingVotes := func(v *VoteBucket) {
+		if v.isUnstaked() {
+			return
+		}
+		if featureCtx.FixContractStakingWeightedVotes {
+			votes.Add(votes, p.calculateVoteWeight(v, false))
+		} else {
+			votes.Add(votes, v.StakedAmount)
+		}
+	}
+	if !featureCtx.CreatePostActionStates {
+		for _, indexer := range indexers {
+			btks, err := indexer.BucketsByCandidate(candidate, height)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to get BucketsByCandidate from contractStakingIndexer")
+			}
+			for _, b := range btks {
+				accountingVotes(b)
+			}
+		}
+		return votes, nil
 	}
-	for _, indexer := range indexers {
-		btks, err := indexer.BucketsByCandidate(candidate, height)
+	for _, cv := range views {
```

Comment:
nit:
the current code works fine since `views` will be an empty slice before the hard-fork flag. but thinking add `if featureCtx.CreatePostActionStates` to make it more clear?